core/trino-main/src/test/java/io/trino/sql/planner/optimizations/TestAddExchangesPlans.java

```diff
@@ -978,6 +990,210 @@ public void testJoinNotExactlyPartitioned()
+                                project(
+                                        exchange(REMOTE, REPARTITION, FIXED_HASH_DISTRIBUTION,
+                                                project(
+                                                        aggregation(ImmutableMap.of("partial_sum", functionCall("sum", ImmutableList.of("nationkey"))),
+                                                                PARTIAL,
+                                                                tableScan("nation", ImmutableMap.of("nationkey", "nationkey")))))))));
+    }
+
+    @Test
+    public void testUnionAllOnPartitionedAndUnpartitionedSources()
+    {
+        assertDistributedPlan(
+                """
+                     SELECT * FROM (SELECT nationkey FROM nation UNION ALL VALUES (1))
+                """,
+                noJoinReordering(),
+                output(
+                        exchange(LOCAL, REPARTITION, FIXED_ARBITRARY_DISTRIBUTION,
+                                values("1"),
+                                exchange(REMOTE, GATHER, SINGLE_DISTRIBUTION,
+                                        tableScan("nation")))));
+    }
+
+    @Test
+    public void testUnionAllOnDifferentCatalogs()
+    {
+        MockConnectorFactory connectorFactory = MockConnectorFactory.builder()
+                .withGetColumns(schemaTableName -> ImmutableList.of(
+                        new ColumnMetadata("nationkey", BigintType.BIGINT)))
+                .withGetTableHandle(((session, schemaTableName) -> new MockConnectorTableHandle(
+                        SchemaTableName.schemaTableName("default", "nation"),
+                        TupleDomain.all(),
+                        Optional.of(ImmutableList.of(new MockConnectorColumnHandle("nationkey", BigintType.BIGINT))))))
+                .withName("mock")
+                .build();
+        getQueryRunner().createCatalog("mock", connectorFactory, ImmutableMap.of());
+
+        // Need to use JOIN as parent of UNION ALL to expose replacing remote exchange with local exchange
+        assertDistributedPlan(
+                """
+                         SELECT * FROM (SELECT nationkey FROM nation UNION ALL SELECT nationkey FROM mock.default.nation), region
+                """,
+                noJoinReordering(),
+                output(
+                        join(INNER, join -> join
+                                .left(exchange(REMOTE, REPARTITION, FIXED_ARBITRARY_DISTRIBUTION,
+                                        tableScan("nation"),
+                                        node(TableScanNode.class)))
+                                .right(
+                                        exchange(LOCAL, GATHER, SINGLE_DISTRIBUTION,
+                                                exchange(REMOTE, REPLICATE, FIXED_BROADCAST_DISTRIBUTION,
+                                                        tableScan("region")))))));
+    }
+
+    @Test
```

Comment:
add a test for `union_all(union_all(tableScan(a), tableScan(b)), tableScan(c))` and
`union_all(tableScan(a), exchange(replicated(b))` even if these cases are not supported yet

Reply:
> `union_all(union_all(tableScan(a), tableScan(b)), tableScan(c))`

done 
>  `union_all(tableScan(a), exchange(replicated(b))`

it is already tested in `testUnionAllAboveBroadcastJoin`